zerver/views/messages.py

```diff
@@ -289,6 +289,38 @@ def by_pm_with(self, query, operand, maybe_negate):
                             column("recipient_id") == narrow_recipient.id))
             return query.where(maybe_negate(cond))
 
+    def by_huddle_with(self, query, operand, maybe_negate):
+        # type: (Query, str, ConditionTransform) -> Query
+        self_recipient = get_recipient(Recipient.PERSONAL, type_id=self.user_profile.id)
+
+        try:
+            narrow_profile = get_user_profile_by_email(operand)
+        except UserProfile.DoesNotExist:
+            raise BadNarrowOperator('unknown user ' + operand)
+        narrow_recipient = get_recipient(Recipient.PERSONAL, narrow_profile.id)
+
+        query = query.select_from(join(query.froms[0], table("zerver_recipient"),
+                                       column("recipient_id") ==
+                                       literal_column("zerver_recipient.id")))
+
+        query_for_self_huddle = select([column("recipient_id")],
+                                       column("user_profile_id") == self_recipient.id,
+                                       table("zerver_subscription"))
+
+        query_for_narrow_huddle = select([column("recipient_id")],
+                                         column("user_profile_id") == narrow_recipient.id,
+                                         table("zerver_subscription"))
+
+        sa_conn = get_sqlalchemy_connection()
+        query_self_result = list(sa_conn.execute(query_for_self_huddle).fetchall())
+        query_narrow_result = list(sa_conn.execute(query_for_narrow_huddle).fetchall())
+
+        cond = and_(column("type") == Recipient.HUDDLE,
+                    column("recipient_id").in_([user_profile_id[0] for user_profile_id in query_self_result]),
+                    column("recipient_id").in_([narrow_profile_id[0] for narrow_profile_id in query_narrow_result]))
```

Comment:
I think it'd be better to move the `column("type") == Recipient.HUDDLE` check into the queries above; that way, the lists of `recipient_id` will be a lot shorter.

Once we do that, we can just take the intersection of the sets from `query_self_result` and `query_narrow_result` in Python, and `cond` can be just something simple like `column("recipient_id").in_(recipient_ids)`.

Finally, we generally don't use `get_sqlalchemy_connection()` in these functions to do pre-processing, instead using the Django ORM.  So I'm imagining something like:

```
self_recipient_ids = [recipient_tuple[0] for recipient_tuple in Subscription.objects.filter(user_profile=self.user_profile, recipient__type=Recipient.HUDDLE).values("recipient_id")]
narrow_recipient_ids = [recipient_tuple[0] for recipient_tuple in Subscription.objects.filter(user_profile=narrow_profile, recipient__type=Recipient.HUDDLE).values("recipient_id")]
recipient_ids = set(self_recipient_ids) & set(narrow_recipient_ids)

cond = column("recipient_id").in_(recipient_ids)
```